chatkit/src/main/java/com/pusher/chatkit/rooms/RoomStateMachine.kt

```diff
@@ -0,0 +1,290 @@
+                    }
+            }
+            .onReady { update -> machine.update(update) }
+    }
+
+    override fun onCancel() {
+        pendingPromise?.cancel()
+    }
+
+}
+
+private class RetryTask(machine: RoomStateMachine) : InternalRoomTask(machine) {
+    override fun invoke() = Unit
+    override fun onCancel() = Unit
+}
+
+private class LoadMoreTask(machine: RoomStateMachine, val messageCount: Int) : InternalRoomTask(machine) {
+    override fun invoke() = Unit
+    override fun onCancel() = Unit
+}
+
+internal sealed class RoomAction {
+    data class LoadRoom(val roomId: Int) : RoomAction()
+    data class LoadMessages(val room: Room) : RoomAction()
+    data class Join(val room: Room) : RoomAction()
+    data class AddMessage(val room: Room, val message: CharSequence) : RoomAction()
+    data class LoadMore(val messageCount: Int) : RoomAction()
+    object Retry : RoomAction()
+}
+
+private typealias Actor = (RoomAction) -> RoomTask
+
+sealed class RoomState(val type: Type) {
+
+    class Initial internal constructor(private val handle: Actor) : RoomState(INITIAL) {
+        fun loadRoom(roomId: Int): RoomTask = handle(RoomAction.LoadRoom(roomId))
+    }
+
+    data class Idle internal constructor(val roomId: Int) : RoomState(IDLE)
+
+    data class NoMembership internal constructor(
+        private val handle: Actor,
+        val room: Room
+    ) : RoomState(NO_MEMBERSHIP) {
+        fun join(): RoomTask = handle(RoomAction.Join(room))
+    }
+
+    data class RoomLoaded internal constructor(
+        private val handle: Actor,
+        val room: Room
+    ) : RoomState(ROOM_LOADED) {
+        fun loadMessages() = handle(RoomAction.LoadMessages(room))
+    }
+
+    data class Loaded internal constructor(
+        private val handle: Actor,
+        val room: Room,
+        val items: List<Item>
+    ) : RoomState(READY) {
+        fun addMessage(text: CharSequence): RoomTask = handle(RoomAction.AddMessage(room, text))
+        @JvmOverloads
+        fun loadMore(messageCount: Int = 10): RoomTask = handle(RoomAction.LoadMore(messageCount))
+    }
+
+    data class Failed internal constructor(private val handle: Actor, val error: Error) : RoomState(FAILED) {
+        fun retry(): RoomTask = handle(RoomAction.Retry)
+    }
+
+    sealed class Item(val type: Type) {
+        abstract val details: Details
+
+        data class Loaded internal constructor(override val details: Details) : Item(Type.LOADED)
+        data class Pending internal constructor(override val details: Details) : Item(Type.PENDING)
+        data class Failed internal constructor(override val details: Details, val error: Error) : Item(Type.FAILED)
+
+        enum class Type {
+            LOADED, PENDING, FAILED
+        }
+
+        data class Details(val userName: CharSequence, val message: CharSequence)
+
+    }
+
+    enum class Type {
+        INITIAL, IDLE, NO_MEMBERSHIP, ROOM_LOADED, READY, FAILED
```

Comment:
What's the difference between all of these states? For example, what's the difference between `IDLE`, `ROOM_LOADED` and `READY`?

Reply:
These map directly to the type of states. Included just for Java compatibility.